### PR TITLE
Make --validate default on and shows how to turn if off

### LINKED ISSUE #10555: Make --validate default on for kubectl
the --validate flag improves error messages for kubectl, for example by pointing out invalid fields.
Why is it off by default?

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #10555. 

cc @erictune @bgrant0607

### CODE REVIEW

**Comment by bgrant0607** [comment]
I think --validate only makes sense for create and replace. Patch doesn't specify full resources, so at minimum required fields may be missing. Other commands either generate or get the resources, and therefore don't need validation.

**Comment by bgrant0607** [comment]
cc @deads2k

**Comment by deads2k** [comment]
Is this part of a larger push for stronger client-side validation?  The current client-side validation only uses swagger to validate types, but since validation rarely becomes more restrictive and our client already transforms between versions (`resource.Builder` decodes to internal and `RunCreate` re-encodes) having our clients run the full validation for a type/operation tuple would be valuable.  The only time it would bite us is loosening validation constraints between skewed client/server versions and that's workaroundable by disabling the check.

**Comment by JanetKuo** [comment]
Does --validate make sens for rolling-update, too?

**Comment by bgrant0607** [comment]
Good point, @JanetKuo. Yes, for rolling-update, also, in the case that the user provides the full resource.

@deads2k Yes, it's part of a push for stronger validation. The main disadvantages right now are latency (an extra roundtrip with the server), bugs/incompleteness of the swagger, which may lead to more false positives, and the question of whether we should validate using the client's API schema or the server's. Due to the disadvantages, I'm hesitant to apply it in cases where it's not as important.

**Comment by JanetKuo** [comment]
@k8s-bot test this please

**Comment by JanetKuo** [comment]
@k8s-bot test this please

File: pkg/api/validation/schema.go [hunk 1]
```diff
@@ -88,6 +88,9 @@ func (s *SwaggerSchema) ValidateBytes(data []byte) error {
 		return fmt.Errorf("kind not set")
 	}
 	allErrs := s.ValidateObject(obj, apiVersion.(string), "", apiVersion.(string)+"."+kind.(string))
+	if len(allErrs) > 0 {
+		glog.Warningf("found schema errors; if you choose to ignore them, turn validation off with --validate=false")
```
Comment (deads2k) [inline pkg/api/validation/schema.go]: Outputing this warning here will not respect the `io.Writer` that has been set for the command and it leaks warning level output from a utility function.  Maybe turn the message into a constant that can be referenced at the point of use in the commands?

File: pkg/kubectl/cmd/util/factory.go [hunk 2]
```diff
@@ -254,8 +254,8 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 			}
 			return kubectl.ReaperFor(mapping.Kind, client, expClient)
 		},
-		Validator: func() (validation.Schema, error) {
-			if flags.Lookup("validate").Value.String() == "true" {
+		Validator: func(validate bool) (validation.Schema, error) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: I vastly prefer this to looking up the flag from inside the function, but why would something with `validate == false` ever call the `Validator()` function?

File: examples/guestbook-go/redis-master-service.json [hunk 3]
```diff
@@ -12,7 +12,8 @@
       "ports": [
         {
           "port":6379,
-          "targetPort":"redis-server"
+          "targetPort":"redis-server",
+          "nodePort":0
```
Comment (deads2k) [inline examples/guestbook-go/redis-master-service.json]: Why this change?

File: pkg/kubectl/cmd/util/factory.go (hunk 2 above)
Comment (JanetKuo) [inline pkg/kubectl/cmd/util/factory.go]: Since `--validate` only makes sense for certain commands, I removed it from global flags, and therefore now this flag is in `cmd.Flags()` instead of `factory.flags`. This means to check the value of this flag, we should use `cmd.Flags().GetBool()` instead of `flags.Lookup()`. Passing the bool value here is better than passing `cmd`, imo. If `Validator()` is called with `validate == false`, `validation.NullSchema{}` will be returned.

File: examples/guestbook-go/redis-master-service.json (hunk 3 above)
Comment (JanetKuo) [inline examples/guestbook-go/redis-master-service.json]: e2e test uses this json file to create services, and I changed it because the validator complains about the schema error (nodePort is required).

File: pkg/kubectl/cmd/util/factory.go (hunk 2 above)
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: > Since --validate only makes sense for certain commands, I removed it from global flags, and therefore now this flag is in cmd.Flags() instead of factory.flags. This means to check the value of this flag, we should use cmd.Flags().GetBool() instead of flags.Lookup(). Passing the bool value here is better than passing cmd, imo. If Validator() is called with validate == false, validation.NullSchema{} will be returned.

I think we talked passed each other.  I agree, this factorization is better than what previously existed.  My question is, if a caller has a value for `shouldValidate`, why wouldn't he do something like
```
if shouldValidate{
    f.Validator()
}
```
?

File: examples/guestbook-go/redis-master-service.json (hunk 3 above)
Comment (deads2k) [inline examples/guestbook-go/redis-master-service.json]: > e2e test uses this json file to create services, and I changed it because the validator complained about the schema error (nodePort is required).

That strikes me as odd.  I thought only certain `ServiceTypes` required `nodePort`.

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (JanetKuo) [inline pkg/api/validation/schema.go]: I agree with what you said, but I used glog warnings because all the validation failure are printed as glog errors, which looks like:
```console
$ cluster/kubectl.sh create -f hello-world3-rc.yaml 
E0819 10:57:37.145443   51079 schema.go:147] Validation failed for: containers, [map[name:hello command:[/bin/echo hello, world] image:ubuntu:14.04 k8s:true]]
E0819 10:57:37.145517   51079 schema.go:147] Validation failed for: spec, map[containers:[map[command:[/bin/echo hello, world] image:ubuntu:14.04 k8s:true name:hello]] restartPolicy:OnFailure]
E0819 10:57:37.145534   51079 schema.go:147] Validation failed for: template, map[metadata:map[labels:map[name:hello-world]] spec:map[restartPolicy:OnFailure containers:[map[command:[/bin/echo hello, world] image:ubuntu:14.04 k8s:true name:hello]]]]
E0819 10:57:37.145556   51079 schema.go:147] Validation failed for: spec, map[template:map[metadata:map[labels:map[name:hello-world]] spec:map[containers:[map[command:[/bin/echo hello, world] image:ubuntu:14.04 k8s:true name:hello]] restartPolicy:OnFailure]] replicas:2 selector:map[name:hello-world]]
W0819 10:57:37.145657   51079 schema.go:92] found schema errors; if you choose to ignore them, turn validation off with --validate=false
error validating "hello-world3-rc.yaml": error validating data: found invalid field k8s for v1.Container
```

Should we change that too?

File: pkg/kubectl/cmd/util/factory.go (hunk 2 above)
Comment (JanetKuo) [inline pkg/kubectl/cmd/util/factory.go]: I see. The caller should only call when they should validate, and we don't need the `validation.NullSchema{}` in Validator.

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (deads2k) [inline pkg/api/validation/schema.go]: > Should we change that too?

Yes, the utility method should avoid output at levels >= `Info`.

File: pkg/kubectl/cmd/util/factory.go (hunk 2 above)
Comment (JanetKuo) [inline pkg/kubectl/cmd/util/factory.go]: Correction: the caller gets `schema` back from `Validator()` and sets the schema of resource builder. Therefore, it makes the code cleaner to always get a schema back, and inside `Validator()` we use `shouldValidate` to determine which schema (swagger or null) to be returned.

File: examples/guestbook-go/redis-master-service.json (hunk 3 above)
Comment (JanetKuo) [inline examples/guestbook-go/redis-master-service.json]: You're right. In [types.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1181) `nodePort` is marked as `json:"nodePort,omitempty"` and should be optional. It's weird that [swagger](http://kubernetes.io/third_party/swagger-ui/#!/api%2Fv1/createNamespacedService) didn't mark it as optional:
```yaml
v1.ServicePort {
  name (string, optional): the name of this port; optional if only one port is defined ,
  protocol (string, optional): the protocol used by this port; must be UDP or TCP; TCP if unspecified ,
  port (integer): the port number that is exposed ,
  targetPort (string, optional): number or name of the port to access on the pods targeted by the service; defaults to the service port; number must be in the range 1 to 65535; name must be an IANA_SVC_NAME; see http://releases.k8s.io/HEAD/docs/services.md#defining-a-service ,
  nodePort (integer): the port on each node on which this service is exposed when type=NodePort or LoadBalancer; usually assigned by the system; if specified, it will be allocated to the service if unused or else creation of the service will fail; see http://releases.k8s.io/HEAD/docs/services.md#type--nodeport
}
```

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (JanetKuo) [inline pkg/api/validation/schema.go]: Fixed.

File: examples/guestbook-go/redis-master-service.json (hunk 3 above)
Comment (bgrant0607) [inline examples/guestbook-go/redis-master-service.json]: This was just changed. Maybe the swagger didn't get updated.

Comment (bgrant0607) [inline examples/guestbook-go/redis-master-service.json]: Which copy of the swagger files does the validator use?

cc @nikhiljindal

Comment (nikhiljindal) [inline examples/guestbook-go/redis-master-service.json]: Yes I see that it was changed recently: https://github.com/kubernetes/kubernetes/pull/12831
http://kubernetes.io/third_party/swagger-ui has not been updated yet, but e2e should be using https://github.com/kubernetes/kubernetes/tree/master/api/swagger-spec which is updated.

**Comment by JanetKuo** [comment]
@k8s-bot test this please

File: examples/guestbook-go/redis-master-service.json (hunk 3 above)
Comment (JanetKuo) [inline examples/guestbook-go/redis-master-service.json]: The failed e2e test ran before this change... now the failure is gone, and I've removed the "nodePort" field from this json file.

**Comment by JanetKuo** [comment]
@deads2k PTAL

File: pkg/kubectl/resource/visitor.go [hunk 4]
```diff
@@ -201,7 +204,7 @@ func ValidateSchema(data []byte, schema validation.Schema) error {
 		return fmt.Errorf("error converting to YAML: %v", err)
 	}
 	if err := schema.ValidateBytes(data); err != nil {
-		return fmt.Errorf("error validating data: %v", err)
+		return fmt.Errorf("error validating data: %v; %s", err, stopValidateMessage)
```
Comment (deads2k) [inline pkg/kubectl/resource/visitor.go]: How about an issue to create an `AggregatedValidationError` type so that we could plumb the handling for this all the way back to the caller?

**Comment by deads2k** [comment]
> How about an issue to create an AggregatedValidationError type so that we could plumb the handling for this all the way back to the caller?

The error type doesn't make this super easy, but as long as you don't wrap the errors here or in the few visitors that call `ValidateSchema`, then you'll be able to test errors that come back from a `Visit` to know if they failed due to validation problems.

I'm ok taking that change as a followup.

lgtm  otherwise.

**Comment by JanetKuo** [comment]
@deads2k I like your idea. The follow-up issue is created.  

@erictune PTAL

**Comment by JanetKuo** [comment]
cc @jlowdermilk

**Comment by erictune** [comment]
I don't think I can add any value by reviewin this on top of what @bgrant0607 and @jlowdermilk and @deads2k  could say.  Assigning to @jlowdermilk .

File: pkg/kubectl/resource/visitor.go [hunk 5]
```diff
@@ -201,7 +204,7 @@ func ValidateSchema(data []byte, schema validation.Schema) error {
 		return fmt.Errorf("error converting to YAML: %v", err)
 	}
 	if err := schema.ValidateBytes(data); err != nil {
```
Comment (jlowdermilk) [inline pkg/kubectl/resource/visitor.go]: What is the behavior of this when schema is `validation.NullSchema{}`

Comment (JanetKuo) [inline pkg/kubectl/resource/visitor.go]: It does nothing but return nil. See https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/schema.go#L53.

**Comment by jlowdermilk** [comment]
lgtm. You may want to fix the commit message (s/trun/turn), but not a big deal.

**Comment by nikhiljindal** [comment]
Needs rebase

**Comment by gmarek** [comment]
@JanetKuo please rebase.
